rs/index/src/ivf/index.rs

```diff
@@ -0,0 +1,97 @@
+use std::collections::HashMap;
+
+use kmeans::*;
+use log::debug;
+use utils::l2::L2DistanceCalculator;
+use utils::DistanceCalculator;
+
+use crate::index::Index;
+
+pub struct Ivf {
+    // The whole dataset. TODO(tyb0807): reduce the memory footprint.
+    dataset: Vec<Vec<f32>>,
+    // Number of clusters.
+    num_clusters: usize,
+    // Each cluster is represented by a centroid vector. This is all the centroids in our IVF.
+    centroids: Vec<Vec<f32>>,
+    // Inverted index mapping each cluster to the vectors it contains.
+    //   key: centroid index in `centroids`
+    //   value: vector index in `dataset`
+    inverted_lists: HashMap<usize, Vec<usize>>,
+    // Number of probed centroids.
+    num_probes: usize,
+}
+
+impl Ivf {
+    pub fn new(
+        dataset: Vec<Vec<f32>>,
+        centroids: Vec<Vec<f32>>,
+        num_clusters: usize,
+        num_probes: usize,
+    ) -> Self {
+        let mut inverted_lists = HashMap::new();
+        // Assign vectors to nearest centroids
+        for (i, vector) in dataset.iter().enumerate() {
+            let nearest_centroid = Self::find_nearest_centroid(&vector, &centroids);
+            inverted_lists
+                .entry(nearest_centroid)
+                .or_insert_with(Vec::new)
+                .push(i);
+        }
+        Self {
+            dataset,
+            num_clusters,
+            centroids,
+            inverted_lists,
+            num_probes,
+        }
+    }
+
+    fn find_nearest_centroid(vector: &Vec<f32>, centroids: &Vec<Vec<f32>>) -> usize {
+        let mut calculator = L2DistanceCalculator::new();
+        centroids
+            .iter()
+            .enumerate()
+            .min_by(|(_, a), (_, b)| {
+                let dist_a = calculator.calculate(&vector, &a);
+                let dist_b = calculator.calculate(&vector, &b);
+                dist_a.partial_cmp(&dist_b).unwrap()
+            })
+            .map(|(index, _)| index)
+            .unwrap()
+    }
+
+    pub fn search(&self, query: &[f32], k: usize) -> Vec<(usize, f32)> {
+        let mut results = Vec::new();
+        let mut probed_centroids = Vec::new();
+
+        // Find the nearest centroids to the query.
+        for _ in 0..self.num_probes {
+            let nearest_centroid = Self::find_nearest_centroid(&query.to_vec(), &self.centroids);
+            if !probed_centroids.contains(&nearest_centroid) {
+                probed_centroids.push(nearest_centroid);
+            }
+        }
+
+        // Search in the inverted lists of the nearest centroids.
+        for &centroid in &probed_centroids {
```

Comment:
to avoid storing the whole list + distance from each embedding to the query vector, use a BinaryHeap here, and then trim them when they reach a threshold (maybe 2*k size)?